cmd/terramate/e2etests/exp_generate_test.go

```diff
@@ -0,0 +1,233 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2etest
+
+import (
+	"fmt"
+	"path/filepath"
+	"testing"
+
+	. "github.com/mineiros-io/terramate/test/hclwrite/hclutils"
+	"github.com/mineiros-io/terramate/test/sandbox"
+)
+
+func TestGenerateDebug(t *testing.T) {
+	type (
+		file struct {
+			path string
+			body fmt.Stringer
+		}
+		testcase struct {
+			name    string
+			layout  []string
+			wd      string
+			configs []file
+			want    runExpected
+		}
+	)
+	t.Parallel()
+
+	testcases := []testcase{
+		{
+			name: "empty project",
+			want: runExpected{},
+		},
+		{
+			name: "stacks with no codegen",
+			layout: []string{
+				"s:stack-1",
+				"s:stack-2",
+			},
+			want: runExpected{},
+		},
+		{
+			name: "stacks with codegen with root as working dir",
+			layout: []string{
+				"s:stack-1",
+				"s:stack-2",
+			},
+			configs: []file{
+				{
+					path: "config.tm",
+					body: Doc(
+						GenerateFile(
+							Labels("file.txt"),
+							Bool("condition", false),
+							Str("content", "data"),
+						),
+						GenerateFile(
+							Labels("file.txt"),
+							Bool("condition", true),
+							Str("content", "data"),
+						),
+					),
+				},
+				{
+					path: "stack-1/config.tm",
+					body: Doc(
+						GenerateHCL(
+							Labels("file.hcl"),
+							Bool("condition", true),
+							Content(
+								Str("content", "data"),
+							),
+						),
+					),
+				},
+				{
+					path: "stack-2/config.tm",
+					body: Doc(
+						GenerateHCL(
+							Labels("file.hcl"),
+							Bool("condition", true),
+							Content(
+								Str("content", "data"),
+							),
+						),
+					),
+				},
+			},
+			want: runExpected{
+				Stdout: `Generated files for /stack-1:
```

Comment:
as this is an experimental feature (and we are in the process of changing outputs of some commands), testing against stdout could be a problem.
Don't need to change, but if a request to change the output comes later, maybe makes sense to just test the data structure (non e2e) and then maybe re-add this test once it becomes a non-exp feature.

Reply:
The in structure tests already exist, it would be the load_test, because all the info/logic is there. This was an extra just to ensure the output is sane, but indeed changing it later will be more work.